### PR TITLE
Let CI distinguish test compilation and test run failures

### PR DESCRIPTION
**Pull request by mherrmann**
Prerequisite for https://github.com/brave/devops/issues/8517.

At the moment, we have a single command for running tests: `npm run test`. This both builds and runs the tests. If the command fails, then it is unclear whether the failure was in building or running the tests. This leads to confusing error messages in our CI. The present PR is a preparation for fixing this.

This PR also incorporates a few small refactorings. Most notably, it introduces internal `buidTests` and `runTests` functions that better separate those concerns.

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by petemill — APPROVED** [review]
package.json changes seem fine. Please get another approval for the code changes before merge. I was only able to cursory review and whilst there are some minor issues (top-level anonymous functions and using let instead of const, that whole file has them and we need to improve the linter to catch these).

File: browser/brave_ads/ads_service_browsertest.cc
Comment (tmancey) [inline browser/brave_ads/ads_service_browsertest.cc:433]: @mherrmann was this change intentional? Thanks

Comment (mherrmann) [reply browser/brave_ads/ads_service_browsertest.cc:433]: @tmancey yes! I am touching the test reporting functionality here. So I purposely introduced a failing test to see whether it still reports failures correctly. It does! I'm reverting the change.

File: build/commands/scripts/commands.js
Comment (bridiver) [inline build/commands/scripts/commands.js:267]: I think this would be a lot simpler if we had an optional `--target` param for the `build` command. Separately we should do the same thing for create_dist

Comment (bridiver) [reply build/commands/scripts/commands.js:267]: for backwards compat you change package.json for `npm run test` to handle this as two commands

Comment (mherrmann) [reply build/commands/scripts/commands.js:267]: > I think this would be a lot simpler if we had an optional --target param for the build

We have such a target, don't we? For [example](https://github.com/brave/brave-core/pull/15440):

    npm run build -- --target=brave/components/policy:pack_policy_templates

I'm pretty sure I'm not fully understanding what you wrote. Could you clarify in more detail what you meant?

PS - as always ty for the review.

Comment (mherrmann) [reply build/commands/scripts/commands.js:267]: > for backwards compat you change package.json for npm run test to handle this as two commands

I'm afraid I don't understand the message of this comment either. Yes I'm changing package.json in a way that's backwards compatible (while also retaining the `test` command for developer ease). Is there something I should change?

Comment (mherrmann) [reply build/commands/scripts/commands.js:267]: @bridiver could I ask you to clarify your comments?

**Review by bridiver — APPROVED** [review]

**Pull request merged**
